blackjax/adaptation/adjusted_mclmc_adaptation.py

```diff
@@ -0,0 +1,378 @@
+            kernel=mclmc_kernel,
+            dim=dim,
+            frac_tune1=frac_tune1,
+            frac_tune2=frac_tune2,
+            target=target,
+            diagonal_preconditioning=diagonal_preconditioning,
+            max=max,
+            tuning_factor=tuning_factor,
+        )(state, params, num_steps, window_key)
+
+    if frac_tune3 != 0:
+        for i in range(num_windows):
+            part2_key = jax.random.fold_in(part2_key, i)
+            part2_key1, part2_key2 = jax.random.split(part2_key, 2)
+
+            state, params = adjusted_mclmc_make_adaptation_L(
+                mclmc_kernel,
+                frac=frac_tune3,
+                Lfactor=0.5,
+                max=max,
+                eigenvector=eigenvector,
+            )(state, params, num_steps, part2_key1)
+
+            (state, params, _) = adjusted_mclmc_make_L_step_size_adaptation(
+                kernel=mclmc_kernel,
+                dim=dim,
+                frac_tune1=frac_tune1,
+                frac_tune2=0,
+                target=target,
+                fix_L_first_da=True,
+                diagonal_preconditioning=diagonal_preconditioning,
+                max=max,
+                tuning_factor=tuning_factor,
+            )(state, params, num_steps, part2_key2)
+
+    return state, params
+
+
+def adjusted_mclmc_make_L_step_size_adaptation(
+    kernel,
+    dim,
+    frac_tune1,
+    frac_tune2,
+    target,
+    diagonal_preconditioning,
+    fix_L_first_da=False,
+    max="avg",
+    tuning_factor=1.0,
+):
+    """Adapts the stepsize and L of the MCLMC kernel. Designed for adjusted MCLMC"""
+
+    def dual_avg_step(fix_L, update_da):
+        """does one step of the dynamics and updates the estimate of the posterior size and optimal stepsize"""
+
+        def step(iteration_state, weight_and_key):
+            mask, rng_key = weight_and_key
+            (
+                previous_state,
+                params,
+                (adaptive_state, step_size_max),
+                previous_weight_and_average,
+            ) = iteration_state
+
+            avg_num_integration_steps = params.L / params.step_size
+
+            state, info = kernel(
+                rng_key=rng_key,
+                state=previous_state,
+                avg_num_integration_steps=avg_num_integration_steps,
+                step_size=params.step_size,
+                sqrt_diag_cov=params.sqrt_diag_cov,
+            )
+
+            # step updating
+            success, state, step_size_max, energy_change = handle_nans(
+                previous_state,
+                state,
+                params.step_size,
+                step_size_max,
+                info.energy,
+            )
+
+            log_step_size, log_step_size_avg, step, avg_error, mu = update_da(
+                adaptive_state, info.acceptance_rate
+            )
+
+            adaptive_state = DualAveragingAdaptationState(
```

Comment:
seems `mask * new_param + (1 - mask) * old_param` is the common pattern, worth to turn it into a lambda function